### ISSUE TITLE
Add format to Record button

### ISSUE THREAD

**Issue by orschiro** [phw/peek#174]
Here is my challenge. 

My default format is GIF.

But sometimes my recordings become too long and GIF won't work.

Then I need to switch to webm or mp4. 

But often I forget to switch back. 

Annoying because I then have to redo the video.

It would help me a lot if the record button would say "Record to GIF". Or "Record to WebM".

What do you think?
![image](https://user-images.githubusercontent.com/170947/29206732-87dd0f46-7e83-11e7-98e0-42f54978855e.png)

**Comment by phw** [comment]
Yes, could be helpful. Maybe even a quick select. My only concern is that it will make the label longer. But since the label autohides anyway this does not matter much.

I am also no opposed to have some kind of quick switching, e.g. by having an arrow dropdown behind the button :)

**Comment by gort818** [comment]
I have an idea for this, what about have a notify event for the record button and have a drop down menu with formats. I will test this out this weekend.

**Comment by gort818** [comment]
@phw @orschiro Here is a quick test, how does it look? 
What do you guys think?
![screenshot from 2017-10-20 22-04-38](https://user-images.githubusercontent.com/9098121/31848258-c85e1a9c-b5e2-11e7-8a86-68f3a5aecf85.png)

**Comment by orschiro** [comment]
@gort818 this looks perfect! How can I test? :-)

**Comment by gort818** [comment]
@orschiro It is in my test here https://github.com/gort818/peek/tree/format-te.. But for far it is just graphical It won't do anything atm, give me some time and I will get it going and let you know when it is ready to test :)

**Comment by orschiro** [comment]
Sounds good. Thanks, Alessandro!

**Comment by gort818** [comment]
@orschiro ok check it out https://github.com/gort818/peek/tree/format-popover build that and let me know what you think.

**Comment by orschiro** [comment]
@gort818 followed the compile from source instructions [1].

But [2] produces an error:

```
orschiro@x230:/tmp/peek/build$ cmake -DCMAKE_INSTALL_PREFIX=/usr/local ..
-- Could NOT find Vala (missing: VALA_EXECUTABLE) 
CMake Error at /usr/share/cmake-3.9/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find Vala (missing: VALA_EXECUTABLE) (Required is at least
  version "0.22")
Call Stack (most recent call first):
  /usr/share/cmake-3.9/Modules/FindPackageHandleStandardArgs.cmake:377 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindVala.cmake:65 (find_package_handle_standard_args)
  CMakeLists.txt:27 (find_package)


-- Configuring incomplete, errors occurred!
See also "/tmp/peek/build/CMakeFiles/CMakeOutput.log".
```

Can you help?

[1] https://github.com/gort818/peek/tree/format-popover#from-source
[2] cmake -DCMAKE_INSTALL_PREFIX=/usr/local ..

**Comment by gort818** [comment]
You need Vala installed.

**Comment by orschiro** [comment]
Thanks! It finished building peek but if I run it [1], I don't get to see your dropdown menu. 

```
orschiro@x230:/tmp/peek/build$ ls -la
total 672
drwxr-xr-x  9 orschiro orschiro   4096 Okt 22 09:57 .
drwxr-xr-x 12 orschiro orschiro   4096 Okt 22 09:57 ..
-rw-r--r--  1 orschiro orschiro  39387 Okt 22 09:57 application.h
-rw-r--r--  1 orschiro orschiro  45297 Okt 22 09:57 application_internal.h
-rw-r--r--  1 orschiro orschiro  38235 Okt 22 09:57 CMakeCache.txt
drwxr-xr-x 36 orschiro orschiro   4096 Okt 22 09:57 CMakeFiles
-rw-r--r--  1 orschiro orschiro   2264 Okt 22 09:57 cmake_install.cmake
-rw-r--r--  1 orschiro orschiro   1196 Okt 22 09:57 cmake_uninstall.cmake
-rw-r--r--  1 orschiro orschiro   3592 Okt 22 09:57 CPackConfig.cmake
-rw-r--r--  1 orschiro orschiro   4026 Okt 22 09:57 CPackSourceConfig.cmake
-rw-r--r--  1 orschiro orschiro    284 Okt 22 09:57 CTestTestfile.cmake
-rw-r--r--  1 orschiro orschiro   2603 Okt 22 09:57 DartConfiguration.tcl
drwxr-xr-x  5 orschiro orschiro   4096 Okt 22 09:57 data
-rw-r--r--  1 orschiro orschiro  78498 Okt 22 09:57 Makefile
-rwxr-xr-x  1 orschiro orschiro 410800 Okt 22 09:57 peek
drwxr-xr-x  3 orschiro orschiro   4096 Okt 22 09:57 po
drwxr-xr-x  6 orschiro orschiro   4096 Okt 22 09:57 src
drwxr-xr-x  3 orschiro orschiro   4096 Okt 22 09:57 Testing
drwxr-xr-x  3 orschiro orschiro   4096 Okt 22 09:57 tests
drwxr-xr-x  2 orschiro orschiro   4096 Okt 22 09:57 ui
```

[1] orschiro@x230:/tmp/peek/build$ ./peek

**Comment by orschiro** [comment]
Hang on. I think I didn't check out the right branch...

Let me try again!

**Comment by orschiro** [comment]
@gort818 here is my feedback [1]. :-)

[1] https://www.useloom.com/share/ea24a461c6bf4eda887e1b2dd598bd91

**Comment by phw** [comment]
Hey, looking cool so far. This is going into the direction I had imagined.

Two notes:

1. I actually would like to have a way to still do a one-click record start. So we should somehow separate the start recording with default format, and format selection. I had thought of using a [https://developer.gnome.org/gtk3/stable/GtkMenuToolButton.html](GtkMenuToolButton) for this.

2. Having the popover to be actual clickable is a bit tricky, due to the content area being set to clickthrough. Have a look at `AppMenu.update_input_shape()`, it already does this for the appmenu, if it is part of the window (https://github.com/phw/peek/blob/master/src/ui/application-window.vala#L404)

Having something like this in `update_input_shape()`  should work:

```
if (pop_format.visible) {
  var pop_format_region = create_region_from_widget (pop_format);
  window_region.union (pop_format);
}
```

**Comment by gort818** [comment]
@orschiro  thanks for the feedback, if you don't mind can you give it another try?
@phw Thanks for the notes, I have updated my branch to include to GtkMenuButton. Everything works on my end. Here are some screenshots:
![screenshot from 2017-10-22 09-17-05](https://user-images.githubusercontent.com/9098121/31863959-7e6cd594-b70a-11e7-944c-f8c119522253.png)
![screenshot from 2017-10-22 09-17-38](https://user-images.githubusercontent.com/9098121/31863960-7e8dd01e-b70a-11e7-87e2-eef3c03a00c7.png)

**Comment by gort818** [comment]
@orschiro  I was curious so I ran ubuntu 17.10 to test for you, so for so good! Take a look!

![screenshot from 2017-10-22 19-21-06](https://user-images.githubusercontent.com/9098121/31865448-9d916f9e-b75e-11e7-8c74-852a181124b6.png)
![test](https://user-images.githubusercontent.com/9098121/31865450-a531de96-b75e-11e7-8711-e8d0d50f6bfd.gif)

**Comment by orschiro** [comment]
@gort818 great! Here [1] are two more feedbacks. :-)

[1] https://www.useloom.com/share/0a980874fbb143439efd18f4d3aa1072

**Comment by gort818** [comment]
@orschiro Thanks so much for the feedback extremely helpful! I'm not sure why the popmenu is not inheriting the default style on Gnome, on cinnamon it works fine, I will investigate that. 

On your second point what do you think about changing the record label when selecting format? For example if I select gif it would say "record as gif" ? Or have check buttons or radio buttons?

**Comment by gort818** [comment]
Ok we are getting somewhere! 
![screenshot from 2017-10-23 23-55-53](https://user-images.githubusercontent.com/9098121/31918759-ec6a1d74-b84d-11e7-966b-2e1e138cb120.png)

**Comment by gort818** [comment]
@orschiro Ok after hours of testing there is only an issue with Ubuntu themes, specifically Ambiance theme! So I checked for users themes and if they are using Ambiance it shows like the photo above. I have a feeling they are still many bugs in the Ambiance and Radiance themes.
Arc and Adwaita look gorgeous,
![peek-test-ada](https://user-images.githubusercontent.com/9098121/31924863-55fabb3c-b839-11e7-8606-371efe317626.gif)
![peek-test-arc](https://user-images.githubusercontent.com/9098121/31924864-561a916e-b839-11e7-9332-9cb1623a409a.gif)

**Comment by orschiro** [comment]
@gort818 feedback time! :-)

https://www.useloom.com/share/e92b8fa19ecb4a04a97efeb6285f8a35

**Comment by gort818** [comment]
@orschiro Thanks once again for the feedback! Unfortunately I cannot seem to do much about the theme, I had to use a work around just to get the popover to match the headerbar colors This is strictly a Ambiance theme issue with Ubuntu by default buttons are white with black text. So anyway take a look and tell me what you think!
I also added changing of the record button label depending on what format you choose.

@phw If you get a chance I would love if you could check it out, I would like to know your opinion as well. https://github.com/gort818/peek/tree/format-popover

**Comment by gort818** [comment]
@orschiro Ok I couldn't help myself I made specific style changes for user with the ambiance theme. I think this is as far as I can go..(Been tearing my hair out!) but we made some progress!

![peek 2017-10-25 20-45](https://user-images.githubusercontent.com/9098121/32034149-aa07eaaa-b9c5-11e7-8934-f15b33cc34a3.gif)

**Comment by orschiro** [comment]
@gort818 just tried it and it looks and feels great! No further remarks from my side. 

Thanks for your work!

**Comment by gort818** [comment]
@orschiro Great to hear!! I will do so more slight changes and issue a pr for @phw to review.

Thanks for all your testing!

**Comment by orschiro** [comment]
@gort818 you're welcome! I am very glad I could help and test. :-)

**Comment by gort818** [comment]
#207

**Comment by gort818** [comment]
@orschiro This has been merged.

**Comment by orschiro** [comment]
@gort818 fantastic! Thank you all for your work. I guess we can close this? :-)

**Comment by phw** [comment]
@orschiro  The development packages should all soon be up-to-date if you want to try it :)

**Comment by orschiro** [comment]
@phw I will definitely give it a try! :-)
